### ISSUE TITLE
Interpolation support for whole job with ability to source variables form local Env Vars

### ISSUE THREAD

**Issue by hatt** [hashicorp/nomad#1185]
With Nomad 0.3.2, the documentation appears a bit vague when it comes to environment variables. I would like to substitute variables in my HCL file from my local environment before the job is submitted to Nomad. Is there a current/recommended way to do this?

My specific use case is to pass the version of build artefacts for Nomad to update my tasks with. Text substitution via Sed or similar seems a bit clunky for this and Terraform already supports local environment variables so it seems possible at least.

**Comment by dadgar** [comment]
Hey,

This is something we are looking forward to bring in a future version of Nomad!

Thanks,
Alex

**Comment by vrenjith** [comment]
@dadgar Any reference to any open issue we can track to know when this will be ready?

**Comment by dadgar** [comment]
@vrenjith I did a cursory glance through the issues and I don't think there is. The timeline for this one is a little far out as we would use HIL (HashiCorp Interpolation Language) but there is work being done to replace that. So we are waiting for its replacement.

**Comment by hatt** [comment]
@dadgar is it worth renaming this issue or otherwise opening another one to track it? Seems like a common enough feature request that people would be searching for it and having somewhere to show it's known and eventually being worked on would help in that regard.

**Comment by dadgar** [comment]
Fair enough!

**Comment by jstoja** [comment]
Hello @dadgar, any news since last september on this one ?

**Comment by dadgar** [comment]
@jstoja Sorry there is no update yet. We have a plan for tackling this and interpolation using HIL which is our interpolation language. However it is a good chunk of work and we have been tackling what we believe to be more impactful things given that jobs can be templated using other existing tools. We very much would like this to be an easy UX once we have HIL though 👍

**Comment by jstoja** [comment]
Great to hear, thank you @dadgar !

**Comment by jstoja** [comment]
Would you recommend confd as a templating tool to make this work with Vault or do you have something else in mind?

**Comment by dadgar** [comment]
@jstoja I would use consul-template but that is just because I am more familiar with it compared to confd.

**Comment by vrenjith** [comment]
In fact this seems to a simple problem to solve. We went ahead and made the
HCL specification itself as a consul-template and do a run of
consul-template before submitting to nomad.

In fact we leveraged some more capabilities of consul-template to have
conditions and include files to avoid repetitions and to standardise job
templates.
-- 
Regards,
Renjith

**Comment by eoliphan** [comment]
@vrenjith how do you handle job specs that themselves contain inline template stanzas?

**Comment by jstoja** [comment]
@vrenjith We're doing the same now. It would just have been nice to have it directly into nomad since it's already dealing with Consul for many things.
The issue we're having right now is that consul-template is waiting each time there is a missing variable, which is painful.

**Comment by vrenjith** [comment]
So we have been using a consul-temaplate config that uses double square
brackets ([[ and ]]) so that the internal templates are left untouched.

Renjith

**Comment by csawyerYumaed** [comment]
I use sed to do this, currently.

**Comment by pmcatominey** [comment]
Shameless plug for [levant](https://github.com/jrasell/levant/).

**Comment by clinta** [comment]
Looking forward to HIL so that I can do some necessary arithmetic on `${NOMAD_ALLOC_INDEX}` which can't be done with template engine that runs prior to submitting the job.

**Comment by daledude** [comment]
Since dadagarg closed my #4270 incorrectly I'll post this here. The constraint documentation specifically states the below. Maybe correct the documentation or it's a bug?

```
value (string: "") - Specifies the value to compare the attribute against using the specified operation. This can be a literal value, another attribute, or any Nomad interpolated values.
``

**Comment by dadgar** [comment]
@daledude The value can be an attribute or metadata on the node not from the jobspec at this moment

**Comment by stevenscg** [comment]
@Mayeu I don't think there has been much movement on this, but I can recommend [levant](https://github.com/jrasell/levant) for templating anything and everything within your job file prior to submission.

**Comment by daledude** [comment]
Don't count on this ever being done (or fixed since the docs say you can do things you cant).

**Comment by vvitayau** [comment]
Hi @csawyerYumaed,

Could you help to clarify how to sed  inside nomad hcl to perform ${NOMAD_ALLOC_INDEX}+1
```
      env {
        ZOOKEEPER_SERVER_ID = "${NOMAD_ALLOC_INDEX}"
        ZOOKEEPER_CLIENT_PORT = 2181
        ZOOKEEPER_SERVERS = "foo.zookeeper.service.consul:2888:3888;bar.zookeeper.service.consul:2888:3888;baz.zookeeper.service.consul:2888:3888"
      }
```

**Comment by csawyerYumaed** [comment]
I'm not sure what you mean about ${NOMAD_ALLOC_INDEX} as that comes from nomad itself..

Here is an example of how we use it:

job.nomad.sed
```
config {
				image = "registry.service.consul:5000/jobname:@@BUILD_NUMBER@@"
				network_mode = "host"
```

and my Makefile that runs from our CI/CD system:
```
deploy: build
	sed -e "s/@@BUILD_NUMBER@@/$(BUILD_NUMBER)/g"  job.nomad.sed >job.nomad
	nomad run job.nomad
```

Both of these are just snippets obviously, but sed just takes values from my env (here the BUILD_NUMBER as defined by my CI/CD system as $BUILD_NUMBER and makes it the docker image tag name in the nomad file).  Before nomad grew up and interacted with vault, it was also how we injected secret values (passwords, TLS certs, etc). Now we just let nomad talk directly to vault for that.

the @@ pre and post fixcharacters  are just convention we use internally, since no self-respecting template language would use such a thing :). Our tools are to busy working to be concerned with self respect :) hehe.

**Comment by cgbaker** [comment]
related: #1185

**Comment by prologic** [comment]
Any updates on progress here? We *could* really utilise this quite well in our Nomad job specs.

**Comment by prologic** [comment]
As an interesting possible work-around for lack of full interpolation support for Nomad Job(s) (*AFAIK coming once HIL is ready*); I came across [Levant](https://github.com/jrasell/levant) which seems to borrow some ideas from Terraform. -- One thing I don't particular like however is that it now mixes two templating languages (*HCL and whatever Levant is*) so it becomes "spaghetti" fast :/

**Comment by stevenscg** [comment]
@prologic Levant really is a great solution for this use case. We also use it for scaling service jobs. The templating syntax is very similar to consul-template and nomad template syntax, so there is at least some consistency if you already use those features.

**Comment by adawalli** [comment]
With @jrasell working at Hashicorp now - here's to hoping that some of the best parts of `levant` can be brought into Nomad Proper. Unfortunately, the current implementation of Nomad is very unfriendly to CI/CD pipelines where Docker Tags might be changing on every single commit. Of course we can use `envsubst` or `levant` but this seems like such a natural thing for Nomad to handle!

**Comment by tgross** [comment]
Cross-linking to https://github.com/hashicorp/nomad/issues/7220

**Comment by tgross** [comment]
HCL2 has shipped in Nomad 1.0, which allows the interpolation of variables from the [job submitter's environment]https://www.nomadproject.io/docs/job-specification/hcl2/variables#environment-variables).

**Comment by Fuco1** [comment]
[Here](https://github.com/hashicorp/nomad/issues/3469#issuecomment-340597208) @dadgar wrote:

> To clarify how the link issue would resolve this, one could set the priority of the job to be a variable supplied as metadata provided when dispatching the parameterized job. Currently it isn't possible since interpolation is client side only and all fields that impact scheduling must be known at job submission time currently.

But I don't see anything about the dispatch meta in the linked hcl2 article.  Does this feature support parametrized job meta arguments?

**Comment by tgross** [comment]
Hi @Fuco1, HCL2 will let you interpolate fairly arbitrary spots in the jobspec. But it's better to open a new issue or post on [Discuss](https://discuss.hashicorp.com/c/nomad/28) than posting responses on a closed issue.
